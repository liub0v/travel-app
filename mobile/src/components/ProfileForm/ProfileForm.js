import {
  Avatar,
  BoldWhiteText,
  GreyText,
  InfoContainer,
  InfoItem,
  MainInfo,
  WhiteText,
} from '../../screens/ProfileScreen/Profile.style';
import React, {useMemo} from 'react';
import {dateParser} from '../../services/dataParser';
import {DEFAULT_PROFILE_IMAGE} from '../../constants/api';
import {AvatarWrapper} from '../../screens/EditProfileScreen/EditProfileScreen.style';
import {AnimatedImage} from '../Loaders/AnimatedImage';
import {ProfileFormWrapper, ProfileInfoWrapper} from './ProfileInfo.style';

export const ProfileForm = ({profileInfo, userInfo, isLoading}) => {
  let birthDate = dateParser(profileInfo?.birthDate);

  const nameTitle = useMemo(() => {
    const firstName = profileInfo?.firstName ?? '';
    const lastName = profileInfo?.lastName ?? '';
    const name = `${firstName} ${lastName}`;
    return name.trim() || userInfo?.username || 'Anonymous';
  }, [profileInfo, userInfo]);

  return (
    <>
      {!isLoading && (
        <ProfileFormWrapper>
          <MainInfo>
            <AvatarWrapper>
              {profileInfo?.imageURL ? (
                <AnimatedImage
                  imageStyle={{width: 125, height: 125, borderRadius: 67.5}}
                  viewStyle={{borderRadius: 67.5}}
                  imageURL={profileInfo?.imageURL}
                />
              ) : (
                <Avatar source={DEFAULT_PROFILE_IMAGE} />
              )}
            </AvatarWrapper>

            {profileInfo ? (
              <>
                <BoldWhiteText>{nameTitle}</BoldWhiteText>
                {profileInfo?.address && (
                  <GreyText>{`${profileInfo?.address}`}</GreyText>
                )}
              </>
            ) : (
              <BoldWhiteText>{`${userInfo?.username}`}</BoldWhiteText>
            )}
          </MainInfo>
          <InfoContainer>
            {userInfo?.username && (
              <InfoItem testID={'username'}>
                <GreyText>{'Username'}</GreyText>
                <WhiteText>{`${userInfo?.username}`}</WhiteText>
              </InfoItem>
            )}
            {userInfo?.email && (
              <InfoItem testID={'email'}>
                <GreyText>{'Email'}</GreyText>
                <WhiteText>{`${userInfo?.email}`}</WhiteText>
              </InfoItem>
            )}
            {profileInfo && (
              <ProfileInfoWrapper>
                {profileInfo?.phone && (
                  <InfoItem testID={'phone'}>
                    <GreyText>{'Phone'}</GreyText>
                    <WhiteText>{`${profileInfo?.phone}`}</WhiteText>
                  </InfoItem>
                )}
                {profileInfo?.birthDate && (
                  <InfoItem testID={'birthDate'}>
                    <GreyText>{'Date of birth'}</GreyText>
                    <WhiteText>{`${birthDate}`}</WhiteText>
                  </InfoItem>
                )}
                {profileInfo?.address && (
                  <InfoItem testID={'address'}>
                    <GreyText>{'Address'}</GreyText>
                    <WhiteText>{`${profileInfo?.address}`}</WhiteText>
                  </InfoItem>
                )}
              </ProfileInfoWrapper>
            )}
          </InfoContainer>
        </ProfileFormWrapper>
      )}
    </>
  );
};
