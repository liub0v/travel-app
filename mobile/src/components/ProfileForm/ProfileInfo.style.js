import styled from 'styled-components/native/dist/styled-components.native.esm';
import fonts from '../../constants/fonts';
import colors from '../../constants/colors';

export const ProfileFormWrapper = styled.View``;
export const ProfileInfoWrapper = styled.View``;
