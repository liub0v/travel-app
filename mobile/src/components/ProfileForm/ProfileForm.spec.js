import React from 'react';
import renderer from 'react-test-renderer';
import {ProfileForm} from './ProfileForm';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {ProfileFormWrapper, ProfileInfoWrapper} from './ProfileInfo.style';
import {BoldWhiteText} from '../../screens/ProfileScreen/Profile.style';
import {USER_MOCKS} from '../../tests/__mocks__/user-mocks';

Enzyme.configure({adapter: new Adapter()});
test('renders correctly', () => {
  const tree = renderer
    .create(
      <ProfileForm
        profileInfo={USER_MOCKS.PROFILE_INFO}
        userInfo={USER_MOCKS.USER_INFO}
        isLoading={false}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Testing ProfileForm component', () => {
  let wrapper;
  let profileInfo;
  let userInfo;

  describe('Testing loading', () => {
    beforeEach(() => {
      wrapper = shallow(
        <ProfileForm
          profileInfo={USER_MOCKS.PROFILE_INFO}
          userInfo={USER_MOCKS.USER_INFO}
        />,
      );
    });
    it('loading is false', () => {
      wrapper = shallow(<ProfileForm isLoading={false} />);
      wrapper.setProps({isLoading: false});
      expect(wrapper.find(ProfileFormWrapper)).toExist();
    });
    it('loading is true', () => {
      wrapper = shallow(<ProfileForm isLoading={true} />);
      wrapper.setProps({isLoading: true});
      expect(wrapper.find(ProfileFormWrapper)).not.toExist();
    });
  });

  describe('Testing fields', () => {
    describe('name field', () => {
      beforeEach(() => {
        profileInfo = {...USER_MOCKS.PROFILE_INFO};
        wrapper = shallow(
          <ProfileForm isLoading={false} userInfo={USER_MOCKS.USER_INFO} />,
        );
      });
      it('only lastName & undefined', () => {
        profileInfo.firstName = undefined;
        wrapper.setProps({profileInfo});
        expect(wrapper.find(BoldWhiteText)).toHaveText(profileInfo.lastName);
      });
      it('only lastName & empty string', () => {
        profileInfo.firstName = '';
        wrapper.setProps({profileInfo});
        expect(wrapper.find(BoldWhiteText)).toHaveText(profileInfo.lastName);
      });
      it('only firstName & undefined', () => {
        profileInfo.lastName = undefined;
        wrapper.setProps({profileInfo});
        expect(wrapper.find(BoldWhiteText)).toHaveText(profileInfo.firstName);
      });
      it('only firstName & empty string', () => {
        profileInfo.lastName = '';
        wrapper.setProps({profileInfo});
        expect(wrapper.find(BoldWhiteText)).toHaveText(profileInfo.firstName);
      });
      it('full name', () => {
        wrapper.setProps({profileInfo});
        expect(wrapper.find(BoldWhiteText)).toHaveText(
          `${profileInfo.firstName} ${profileInfo.lastName}`,
        );
      });
    });
    describe('user info fields', () => {
      beforeEach(() => {
        userInfo = {...USER_MOCKS.USER_INFO};
        wrapper = shallow(
          <ProfileForm
            isLoading={false}
            profileInfo={USER_MOCKS.PROFILE_INFO}
          />,
        );
      });
      it('userInfo is undefined', () => {
        userInfo = undefined;
        wrapper.setProps({userInfo});
        expect(wrapper.find({testID: 'username'})).not.toExist();
        expect(wrapper.find({testID: 'email'})).not.toExist();
      });

      it('username is undefined', () => {
        userInfo.username = undefined;
        wrapper.setProps({userInfo});
        expect(wrapper.find({testID: 'username'})).not.toExist();
      });
      it('email is undefined', () => {
        userInfo.email = undefined;
        wrapper.setProps({userInfo});
        expect(wrapper.find({testID: 'email'})).not.toExist();
      });
    });

    describe('if profile info fields are undefined', () => {
      beforeEach(() => {
        profileInfo = {...USER_MOCKS.PROFILE_INFO};
        wrapper = shallow(
          <ProfileForm isLoading={false} userInfo={USER_MOCKS.USER_INFO} />,
        );
      });
      it('profile is undefined', () => {
        profileInfo = undefined;
        wrapper.setProps({profileInfo});
        expect(wrapper.find(ProfileInfoWrapper)).not.toExist();
      });
      it('phone is undefined', () => {
        profileInfo.phone = undefined;
        wrapper.setProps({userInfo});
        expect(wrapper.find({testID: 'phone'})).not.toExist();
      });
      it('birthDate is undefined', () => {
        profileInfo.birthDate = undefined;
        wrapper.setProps({userInfo});
        expect(wrapper.find({testID: 'birthDate'})).not.toExist();
      });
      it('address is undefined', () => {
        profileInfo.address = undefined;
        wrapper.setProps({userInfo});
        expect(wrapper.find({testID: 'address'})).not.toExist();
      });
    });
  });
});
