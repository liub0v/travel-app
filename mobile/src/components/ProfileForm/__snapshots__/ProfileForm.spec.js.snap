// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ProfileForm component Testing loading loading is false 1`] = `ShallowWrapper {}`;

exports[`Testing ProfileForm component Testing loading loading is false 2`] = `ShallowWrapper {}`;

exports[`Testing ProfileForm component Testing loading loading is true 1`] = `ShallowWrapper {}`;

exports[`Testing ProfileForm component Testing loading loading is true 2`] = `ShallowWrapper {}`;

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF7F",
            "borderRadius": 67.5,
            "height": 125,
            "justifyContent": "center",
            "marginBottom": 16,
            "marginTop": 24,
            "width": 125,
          },
        ]
      }
    >
      <View
        collapsable={false}
        nativeID="animatedComponent"
        style={
          Object {
            "backgroundColor": "#FFFFFF7F",
            "borderRadius": 67.5,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 67.5,
                "height": 125,
                "width": 125,
              },
            ]
          }
        >
          <FastImageView
            blurRadius={5}
            onFastImageLoadEnd={[Function]}
            onFastImageLoadStart={[Function]}
            resizeMode={null}
            source={
              Object {
                "uri": "https://cdn.pixabay.com/photo/2016/08/08/09/17/avatar-1577909_960_720.png",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontFamily": "MontserratExtraBold",
            "fontSize": 16,
            "marginBottom": 4,
          },
        ]
      }
    >
      FirstName LastName
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Montserrat",
            "fontSize": 14,
            "opacity": 0.5,
          },
        ]
      }
    >
      City, Country
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "rgba(255, 255, 255, 0.2)",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginLeft": 10,
            "marginRight": 10,
            "paddingBottom": 22,
            "paddingTop": 22,
          },
        ]
      }
      testID="username"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Montserrat",
              "fontSize": 14,
              "opacity": 0.5,
            },
          ]
        }
      >
        Username
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Montserrat",
              "fontSize": 14,
            },
          ]
        }
      >
        username
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "rgba(255, 255, 255, 0.2)",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginLeft": 10,
            "marginRight": 10,
            "paddingBottom": 22,
            "paddingTop": 22,
          },
        ]
      }
      testID="email"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Montserrat",
              "fontSize": 14,
              "opacity": 0.5,
            },
          ]
        }
      >
        Email
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Montserrat",
              "fontSize": 14,
            },
          ]
        }
      >
        user@mail.com
      </Text>
    </View>
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "rgba(255, 255, 255, 0.2)",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginLeft": 10,
              "marginRight": 10,
              "paddingBottom": 22,
              "paddingTop": 22,
            },
          ]
        }
        testID="phone"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 14,
                "opacity": 0.5,
              },
            ]
          }
        >
          Phone
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 14,
              },
            ]
          }
        >
          80291234567
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "rgba(255, 255, 255, 0.2)",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginLeft": 10,
              "marginRight": 10,
              "paddingBottom": 22,
              "paddingTop": 22,
            },
          ]
        }
        testID="birthDate"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 14,
                "opacity": 0.5,
              },
            ]
          }
        >
          Date of birth
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 14,
              },
            ]
          }
        >
          November 23, 2001
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "rgba(255, 255, 255, 0.2)",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginLeft": 10,
              "marginRight": 10,
              "paddingBottom": 22,
              "paddingTop": 22,
            },
          ]
        }
        testID="address"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 14,
                "opacity": 0.5,
              },
            ]
          }
        >
          Address
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat",
                "fontSize": 14,
              },
            ]
          }
        >
          City, Country
        </Text>
      </View>
    </View>
  </View>
</View>
`;
